core/src/main/java/org/apache/iceberg/BaseFile.java

```diff
@@ -0,0 +1,389 @@
+      this.nullValueCounts = copy(toCopy.nullValueCounts);
+      this.lowerBounds = SerializableByteBufferMap.wrap(copy(toCopy.lowerBounds));
+      this.upperBounds = SerializableByteBufferMap.wrap(copy(toCopy.upperBounds));
+    } else {
+      this.columnSizes = null;
+      this.valueCounts = null;
+      this.nullValueCounts = null;
+      this.lowerBounds = null;
+      this.upperBounds = null;
+    }
+    this.fromProjectionPos = toCopy.fromProjectionPos;
+    this.keyMetadata = toCopy.keyMetadata == null ? null : Arrays.copyOf(toCopy.keyMetadata, toCopy.keyMetadata.length);
+    this.splitOffsets = copy(toCopy.splitOffsets);
+  }
+
+  /**
+   * Constructor for Java serialization.
+   */
+  BaseFile() {
+  }
+
+  @Override
+  public org.apache.avro.Schema getSchema() {
+    if (avroSchema == null) {
+      this.avroSchema = getAvroSchema(partitionType);
+    }
+    return avroSchema;
+  }
+
+  @Override
+  @SuppressWarnings("unchecked")
+  public void put(int i, Object value) {
+    int pos = i;
+    // if the schema was projected, map the incoming ordinal to the expected one
+    if (fromProjectionPos != null) {
+      pos = fromProjectionPos[i];
+    }
+    switch (pos) {
+      case 0:
+        this.content = value != null ? FileContent.values()[(Integer) value] : FileContent.DATA;
+        return;
+      case 1:
+        // always coerce to String for Serializable
+        this.filePath = value.toString();
+        return;
+      case 2:
+        this.format = FileFormat.valueOf(value.toString());
+        return;
+      case 3:
+        this.partitionData = (PartitionData) value;
+        return;
+      case 4:
+        this.recordCount = (Long) value;
+        return;
+      case 5:
+        this.fileSizeInBytes = (Long) value;
+        return;
+      case 6:
+        this.columnSizes = (Map<Integer, Long>) value;
+        return;
+      case 7:
+        this.valueCounts = (Map<Integer, Long>) value;
+        return;
+      case 8:
+        this.nullValueCounts = (Map<Integer, Long>) value;
+        return;
+      case 9:
+        this.lowerBounds = SerializableByteBufferMap.wrap((Map<Integer, ByteBuffer>) value);
+        return;
+      case 10:
+        this.upperBounds = SerializableByteBufferMap.wrap((Map<Integer, ByteBuffer>) value);
+        return;
+      case 11:
+        this.keyMetadata = ByteBuffers.toByteArray((ByteBuffer) value);
+        return;
+      case 12:
+        this.splitOffsets = (List<Long>) value;
+        return;
+      default:
+        // ignore the object, it must be from a newer version of the format
+    }
+  }
+
+  @Override
+  public <T> void set(int pos, T value) {
+    put(pos, value);
+  }
+
+  @Override
+  public Object get(int i) {
+    int pos = i;
+    // if the schema was projected, map the incoming ordinal to the expected one
+    if (fromProjectionPos != null) {
+      pos = fromProjectionPos[i];
+    }
+    switch (pos) {
+      case 0:
+        return FileContent.DATA.id();
```

Comment:
I guess we should return `this.content.id()` here ?

Reply:
This is only used on read to get a container to reuse, so it doesn't affect correctness. Writes all go through the wrappers, which is why tests pass. I'll fix this to avoid confusion.